packages/react-core/src/core/providers/thirdweb-wallet-provider.tsx

```diff
@@ -0,0 +1,293 @@
+        "lastConnectedWallet",
+      );
+
+      const Wallet = lastConnectedWallet
+        ? props.supportedWallets.find((W) => {
+            console.log("lastConnectedWallet", lastConnectedWallet);
+            console.log("W.name", W.name);
+
+            return W.name
+              .toLowerCase()
+              .includes(lastConnectedWallet?.toLowerCase() || "");
+          })
+        : undefined;
+
+      console.log("found wallet", Wallet);
+
+      if (Wallet && Wallet.id !== "deviceWallet") {
+        const wallet = createWalletInstance(Wallet);
+        try {
+          setConnectionStatus("connecting");
+          // give up auto connect if it takes more than 3 seconds
+          // this is to handle the edge case when trying to auto-connect to wallet that does not exist anymore (extension is uninstalled)
+          await timeoutPromise(
+            3000,
+            wallet.autoConnect(),
+            `AutoConnect timeout`,
+          );
+          handleWalletConnect(wallet);
+        } catch (e) {
+          setConnectionStatus("disconnected");
+          throw e;
+        }
+      } else {
+        setConnectionStatus("disconnected");
+      }
+    })();
+  }, [
+    createWalletInstance,
+    props.supportedWallets,
+    handleWalletConnect,
+    props.shouldAutoConnect,
+    activeWallet,
+    connectionStatus,
+  ]);
+
+  const connectWallet = useCallback(
+    async <W extends SupportedWallet>(
+      Wallet: W,
+      connectParams: Parameters<InstanceType<W>["connect"]>[0],
+    ) => {
+      const _connectedParams = {
+        chainId: chainIdToConnect,
+        ...connectParams,
+      };
+
+      const wallet = createWalletInstance(Wallet);
+      setConnectionStatus("connecting");
+      try {
+        await wallet.connect(_connectedParams);
+        handleWalletConnect(wallet);
+      } catch (e: any) {
+        console.error(e);
+        setConnectionStatus("disconnected");
+        throw e;
+      }
+    },
+    [createWalletInstance, handleWalletConnect, chainIdToConnect],
+  );
+
+  const onWalletDisconnect = useCallback(() => {
+    setConnectionStatus("disconnected");
+    setSigner(undefined);
+    setActiveChainId(undefined);
+    setActiveWallet(undefined);
+  }, []);
+
+  const disconnectWallet = useCallback(async () => {
+    // get the connected wallet
+    if (!activeWallet) {
+      return Promise.resolve();
+    }
+
+    return activeWallet.disconnect().then(() => {
+      onWalletDisconnect();
+    });
+  }, [activeWallet, onWalletDisconnect]);
+
+  // when wallet's network or account is changed using the extension, update UI
+  useEffect(() => {
+    if (!activeWallet) {
+      return;
+    }
+
+    const update = async () => {
+      const _signer = await activeWallet.getSigner();
+      const _chainId = await _signer.getChainId();
+      setSigner(_signer);
+      setActiveChainId(_chainId);
+    };
+
+    activeWallet.addListener("change", () => {
+      update();
```

Comment:
this can technically fire multiple times queing up `update` calls and potentially leading to incorrect states (second update could resolve before first one)

it's probably a non-issue but calling out that it *could* happen

Reply:
You just need to remove the listener in a returning function in the useEffect()